wasm.rkt

```diff
@@ -69,6 +80,80 @@
                                                   0
                                                   1))])
 
+;;;;;;;;; TABLE
+
+;; add table
+(define-metafunction WASM-eval
+  table-add : s tab -> s
+  [(table-add (s-func s-table s-mem) (table c_1 table-type )) (s-func (table-type c_1 (const 0) (const 0) $nopointer s-table) s-mem)])
+
+;;update table offset
+(define-metafunction WASM-eval
+  table-update-offset : s-func s-mem s-table v -> s
+  [(table-update-offset s-func s-mem (table-type c_size v_offset v_index i_1 s-table)  v_new)
+   (s-func  (table-type c_size v_new v_index i_1 mt-s-table) s-mem)])
+
+;;set table offset
+(define-metafunction WASM-eval
+  table-set-offset : s v -> s
+  [(table-set-offset (s-func s-table s-mem)  v_offset) (table-update-offset s-func s-mem s-table v_offset)])
+
+;;Add-update table 
+(define-metafunction WASM-eval
+  table-add-update-function : s-func s-mem s-table i -> s
+  [(table-add-update-function s-func s-mem (table-type c_size v_offset v_index i_func s-table)  i_funcN)
+   ,(if (equal? (term i_func) (term $nopointer))
+        ;;update the table with the function and build up the store
+        (term ( s-func
+                (table-type c_size v_offset v_index i_funcN mt-s-table)
+                s-mem))
+        ;;add the function 
+        (if (< (term (extract-const-value v_offset)) (- (term c_size) 1))
+            ;;add the function to the table and build up the store
+            (term ( s-func
+                    (table-type c_size  (const ,(+ (term (extract-const-value v_offset)) 1))
+                                (const ,(+ (term (extract-const-value v_index)) 1)) i_funcN
+                                (table-type c_size v_offset v_index i_func s-table))
+                    s-mem))
+
+            ;;throw exception
+            (term (func DEBUG-BAD (debug "Table index out of bound" c_size)))))])
+
+;;helper meta-function to extract number from  (const c) value
+(define-metafunction WASM-eval
+  extract-const-value : v -> c
+  [(extract-const-value (const c_1)) c_1])
+
+
+;;set table add function
+(define-metafunction WASM-eval
+  table-operation-function : s i -> s
+  [(table-operation-function (s-func s-table s-mem)  i_funcN) (table-add-update-function s-func s-mem s-table i_funcN)])
+
+;;fetch function from table
+(define-metafunction WASM-eval
+  fetch-from-table : s v -> f
+  [(fetch-from-table (s-func s-table s-mem) v_1) (search-table (s-func s-table s-mem) s-table v_1)])
```

Comment:
instead of 
`[(fetch-from-table (s-func s-table s-mem) v_1) (search-table (s-func s-table s-mem) s-table v_1)])`
could we use
`[(fetch-from-table s (search-table s s-table v_1)])`